Draft: Typebind & autobind using Pi-like syntax (#3120)
idris-lang/Idris2: docs/source/reference/operators.rst

     ...

   distrib1 : {x, y, z : a} -> x + y + z -> (x + y) + z

Here ``distrib1`` makes explicit use of the operator being defined as
right-associative.

Typebind Operators
==================

In dependently-typed programming, we have the ability define types which
first argument is a type and the second is a lambda using the first argument
as it's type. A typical example of this is the dependent linear arrow:

 .. code-block:: idris

    infixr 0 =@
    0 (=@) : (x : Type) -> (x -> Type) -> Type
    (=@) x f = (1 v : x) -> f v

However, we cannot use as is because the second argument is
a lambda, and writing out any value using this operator will look a bit awkward:

.. code-block:: idris

   linearSingleton : Nat =@ (\x => Singleton x)
   linearSingleton = ...

What we really want to write, is something like the dependent arrow ``->`` but
for linear types:

.. code-block:: idris

   linearSingleton : (x : Nat) =@ Singleton x
   linearSingleton = ...

The above syntax is allowed if the operator is declared as ``typebind``. For
this, simply add the ``typebind`` keyword in front of the fixity declaration.

.. code-block:: idris

   typebind infixr 0 =@

> ``typebind`` is a mofidier like ``export`` and both can be used at the same time.

Reviewer: (typo `mofidier`)

Author: thnaks

This new syntax is purely syntax sugar and converts any instance of
``(name : type) op expr`` into ``type op (\name : type => expr)``

Because of its left-to-right binding structure, typebind operators can
only ever be ``infixr`` with precedence 0.

Autobind Operators
==================

Typebind operators allow to bind a *type* on the left side of an operator, but
sometimes, there is no dependency between the first argument, and the expression
on the right side of the operator. For those case, we use ``autobind``.

An example of this is a DSL for a dependently-typed programming language
where the constructor for ``Pi`` takes terms on the left and lambdas on the right:

.. code-block:: idris

    VPi : Value -> (Value -> Value) -> Value

    sig : Value
    sig = VPi VStar                 (\fstTy -> VPi
          (VPi fstTy (const VStar)) (\sndTy -> VPi
          fstTy                     (\val1 -> VPi
          (sndTy `vapp` val1)       (\val2 ->
          VSigma fstTy sndTy)))))

We would like to use a custom operator to build values using ``VPi``, but its
signature does not follow the pattern that ``typebind`` uses. Instead, we use

> ``autobind`` to tell the compiler that the type of the lambda is not given
> by the first argument. For this we use ``:=`` instead of ``:``:

Author: ```suggestion
``autobind`` to tell the compiler that the type of the lambda must be inferred. 
For this we use ``:=`` instead of ``:``:
```
